File: sort/selection.cpp
#include <iostream>

using namespace std;

int partition(int a[], int l, int r) {
  int p=l;
  int v=a[r];
  for (int i=l; i<=r; i++) {
    if (a[i]<v) {
      swap(a[i], a[p]);
      p++;
    }
  }
  swap(a[p], a[r]);
  return p;
}

int selection(int a[], int n, int k) {
  int l, r;
  l = 0;
  r = n-1;
  while(r>=l) {
    int idx = partition(a, l, r);
    if (idx == k) return a[k];
    else if (idx>k) r = idx-1;
    else l = idx+1;
  }
  return 0;
}

int main() {
  int k=0;
  int a[] = {1,6,5,4,3,7};
  while (cin>>k && k!= -1) {
    cout << selection(a, sizeof(a)/sizeof(a[0]), k) << endl;
  }
  return 1;
}

File: sort/Selection.java
import java.util.*;

public class Selection {
    public static void swap(int[] a, int i, int j) {
        if (i==j) return;
        int t = a[i];
        a[i] = a[j];
        a[j] = t;
    }

    public static int partition(int[] a, int left, int right) {
        if (left >= right) return left;
        // select last one as pivot
        int p = a[right]; 
        int i, j;
        i = left;
        j = right-1;
        while (true) {
            while (a[i]<p) i++; //pay attention to this
            while (a[j]>p && i<j) j--;
            if (i>=j) break;
            swap(a, i,j);
        }
        swap(a, i, right);
        return i;
    }

    public static int select(int[] a, int k) {
        if (k<=0 || k > a.length) throw new IllegalArgumentException();
        int l = 0;
        int r = a.length-1;
        while (l <= r) {
            int m = partition(a, l, r);
            if (m+1 == k) return a[m];
            if (m+1 < k) l = m+1;
            else r = m-1;
        }
        return -1;
    }

    public static void main(String[] args) {
        int[] a = {1,6,5,4,3,7};
        Scanner sc = new Scanner(System.in);
        int k;
        while ((k = sc.nextInt()) >= 0) {
            System.out.println(select(a, k));
        }
    }
}
